@azure/communication-common Azure Core 2.0 Migration

petrsvihlik opened the pull request:
### Packages impacted by this PR
- @azure/communication-common
- and once migrated, transitively its dependencies

### Issues associated with this PR

- #20056 
- #15594

### Describe the problem that is addressed by this PR
- https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-rest-pipeline/documentation/core2.md

### What are the possible designs available to address the problem? If there are more than one possible design, why was the one in this PR chosen?
- https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-rest-pipeline/documentation/core2.md

### Are there test cases added in this PR?
- [x] Yes

### Provide a list of related PRs _(if any)_
- #19043
- #17045

### Checklists
- [x] Added impacted package name to the issue description
- [x] Does this PR needs any fixes in the SDK Generator? **NO**
- [x] Added a changelog (if necessary)


### Changes
- old policy preserved to maintain the backward compatibility and allow gradual migration of the ACS libraries and moved to `/src/credential/auth-policy-v1/`
- new policy created in `/src/credential/auth-policy-v2/` (based on `@azure/core-rest-pipeline`)
- reduced the number of references to `@azure/core-http` to minimum (enable easier future removal of the dependency)
- removed circular dependencies (detected by `npx madge --circular --extensions ts ./`)
- removed `SubtleCrypto` & related shims

### Migration guide for ACS modalities
- `createCommunicationAuthPolicy` -> `createCommunicationAuthenticationPolicy`
- `createCommunicationAuthenticationPolicy` -> `createCommunicationAuthenticationPolicy`

```typescript
// Old code
const authPolicy = createCommunicationAuthPolicy(credential);
const pipeline = createPipelineFromOptions(internalPipelineOptions, authPolicy);
this.client = new IdentityRestClient(url, pipeline).communicationIdentity;
```

```typescript
// New code
this.client = new IdentityRestClient(url, { endpoint: url, ...internalPipelineOptions });
const authPolicy = createCommunicationAuthenticationPolicy(credential);
this.client.pipeline.addPolicy(authPolicy);
```

azure-sdk commented:
API changes have been detected in `@azure/communication-common`. You can review API changes [here](https://apiview.dev/Assemblies/Review/b0a054c7d9674778b508a925f4ed1519?diffOnly=True&diffRevisionId=4766e5b3b88041948d84fb1cfefcdf1a)

**API changes**
```diff
+ createCommunicationAuthenticationPolicy: (credential: KeyCredential | TokenCredential) => PipelinePolicy
+ export declare function createCommunicationKeyCredentialPolicy(credential: KeyCredential): PipelinePolicy;
```

azure-sdk commented:
API changes have been detected in `@azure/communication-common`. You can review API changes [here](https://apiview.dev/Assemblies/Review/b0a054c7d9674778b508a925f4ed1519?diffOnly=True&diffRevisionId=4766e5b3b88041948d84fb1cfefcdf1a)

**API changes**
```diff
+ export declare function createCommunicationAuthenticationPolicy(credential: KeyCredential | TokenCredential): PipelinePolicy;
+ export declare function createCommunicationKeyCredentialPolicy(credential: KeyCredential): PipelinePolicy;
```

0rland0Wats0n reviewed (approved):
LGTM @danielav7 can you guys take a look at this as well

xirzec commented on sdk/communication/communication-common/src/credential/url.browser.ts:
```diff
@@ -0,0 +1,9 @@
+const url = URL;
+const urlSearchParams = URLSearchParams;
+
+export { url as URL, urlSearchParams as URLSearchParams };
```
you should be able to use `URL` without these helper files since it is supported by our min node runtime

petrsvihlik replied:
@xirzec well spotted, thanks! it's fixed now :)

petrsvihlik commented:
The PR contains a new version of the auth policy that is used by several ACS modalities (SMS, ShortCodes, NetworkTraversal, Identity, PhoneNumbers). To preserve the backward compatibility and allow gradual onboarding by the aforementioned modalities, I'm keeping both the old and the new version of the policy. To avoid naming collisions while not having to suffix the new one with sth like "V2", I've decided to use slightly different names for the new policy:
	
- `createCommunicationAuthPolicy` -> `createCommunicationAuthenticationPolicy`
- `createCommunicationAuthenticationPolicy `-> `createCommunicationAuthenticationPolicy`

Also, all these methods are marked as @hidden to be excluded from the docs, and I marked the old ones with @deprecated.

I was considering removing the old policy entirely since it's marked as `@hidden` and not supposed to be used by customers but in the end, I decided not to because I couldn't find any Azure guidelines that would allow me to do so.

@xirzec @bterlson Could you please let me know what approach you would recommend and if it's the current one, then approve the related [API view](https://github.com/Azure/azure-sdk-for-js/pull/20337#issuecomment-1039042878)?

xirzec commented:
> @xirzec @bterlson Could you please let me know what approach you would recommend and if it's the current one, then approve the related [API view](https://github.com/Azure/azure-sdk-for-js/pull/20337#issuecomment-1039042878)?

Our normal approach in these sorts of situations (we're updating a core package) is to bump the major version and do the break (so ditch the old policies and export the new ones with the same name) and then have dependent packages upgrade incrementally.

ramya-rao-a commented:
> Both have their pros and cons. If we delete the old one, we should migrate all the client libs in a single PR...a bit more work but less code to maintain in the end. If we stick with the current approach, we allow for gradual onboarding and still can decide to remove it later, once everyone's onboarded.

I agree with @xirzec.
By doing a major version update for communication-common, we will be free to remove dependency on core-http. The rest of the communication libraries can move to this new major version in different PRs.

petrsvihlik commented:
@xirzec I have removed the old policy and replaced it with a new one. No APIView has been generated, I suspect it's because I have bumped the major version as you suggested. Can you please give a final approval so that we can merge this PR? The API change is visible here: https://github.com/Azure/azure-sdk-for-js/pull/20337/files#diff-aa1c3b86239992f00aa0ef7192f0f24632b66e2a0b086a001189b2670cba75e7

xirzec reviewed (approved):
API changes look good! Thanks for the update! 👍

ramya-rao-a commented on sdk/communication/communication-common/CHANGELOG.md:
```diff
@@ -1,10 +1,11 @@
 ### Features Added
 
 - Optimization added: When the proactive refreshing is enabled and the token refresher fails to provide a token that's not about to expire soon, the subsequent refresh attempts will be scheduled for when the token reaches half of its remaining lifetime until a token with long enough validity (>10 minutes) is obtained.
+- Migrated from using `@azure/core-http` to `@azure/core-rest-pipeline` for the handling of HTTP requests. See [Azure Core v1 vs v2](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-rest-pipeline/documentation/core2.md) for more on the difference and benefits of the move.
```
This should go under the breaking changes section with a note on the policies/classes being dropped/replaced

petrsvihlik replied:
thank you, fixed

petrsvihlik commented:
@xirzec in the end, I managed to create an API view for this PR manually. can you please approve it there as well?
https://apiview.dev/Assemblies/Review/cd646a2eb52342bd84aa3d1e3fb4285e/7e0a3470c9584f4cbacba31220aa2889?diffRevisionId=f86bf33908884a1bb597730856d29199&doc=False&diffOnly=True

azure-sdk commented:
API changes have been detected in `@azure/communication-common`. You can review API changes [here](https://apiview.dev/Assemblies/Review/b0a054c7d9674778b508a925f4ed1519?diffOnly=True&diffRevisionId=4766e5b3b88041948d84fb1cfefcdf1a)

**API changes**
```diff
- createCommunicationAccessKeyCredentialPolicy: (credential: KeyCredential) => RequestPolicyFactory
- createCommunicationAuthPolicy: (credential: KeyCredential | TokenCredential) => RequestPolicyFactory
+ export declare function createCommunicationAccessKeyCredentialPolicy(credential: KeyCredential): PipelinePolicy;
+ export declare function createCommunicationAuthPolicy(credential: KeyCredential | TokenCredential): PipelinePolicy;
```

xirzec commented:
> @xirzec in the end, I managed to create an API view for this PR manually. can you please approve it there as well? https://apiview.dev/Assemblies/Review/cd646a2eb52342bd84aa3d1e3fb4285e/7e0a3470c9584f4cbacba31220aa2889?diffRevisionId=f86bf33908884a1bb597730856d29199&doc=False&diffOnly=True
> 
> **edit**:
> 
> * oh, the PR API views started working again.... so you can also approve this one: https://apiview.dev/Assemblies/Review/b0a054c7d9674778b508a925f4ed1519?diffOnly=True&diffRevisionId=4766e5b3b88041948d84fb1cfefcdf1a sorry for the confusion😁

I approved both! so you're super approved. 😆

ramya-rao-a commented:
Lets go! Lets go!
